src/flint/fmpz_factor.jl

```diff
@@ -18,30 +18,45 @@ function _factor(a::ZZRingElem)
 end
 
 # Just to give a helpful error message if someone tries to factor a boolean
-function factor(b::Bool)
-  throw(DomainError("Cannot factorize a boolean"));
+function factor(::Bool)
+  throw(DomainError("Cannot factorize a boolean"))
 end
 
-# This function handles machine integer types (up to 64 bits)
-function factor(a::T) where T <: Union{Int, UInt}
+function factor(::Type{Bool}, ::Int)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+function factor(::Type{Bool}, ::ZZRingElem)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+
+# The main dispatch function: flintify gives either Int or ZZRingElem,
+# then use Julia's dispatcher to select the worker function.
+function factor(a::T)  where T <: Integer
+  iszero(a) && throw(ArgumentError("Argument must be non-zero"))
+  return factor(typeof(a), Nemo.flintify(a))
+end
+
+# Two worker functions: one for Int, one for ZZRingElem
+function factor(::Type{T}, a::Int)  where T <: Int
   iszero(a) && throw(ArgumentError("Argument must be non-zero"))
-  u = sign(a)
-  a = u < 0 ? -a : a
+  u = T(sign(a))
+  abs_a = reinterpret(UInt, a < 0 ? -a : a) # supposedly correct when a is most negative Int value
   F = n_factor()
-  @ccall libflint.n_factor(F::Ref{n_factor}, a::UInt)::Nothing
+  @ccall libflint.n_factor(F::Ref{n_factor}, abs_a::UInt)::Nothing
   res = Dict{T, Int}()  # factor-multiplicity pairs
   for i in 1:F.num
-    z = F.p[i]
+    z = T(F.p[i])
```

Comment:
Is the cast necessary? I think it is automatic handled by `setindex!` below, no?
```suggestion
    z = F.p[i]
```
But of course it also doesn't hurt

Reply:
I prefer to make the cast explicit so that a reader understands that a cast is intentionally occurring.